Makes fire colorful and anime

DaedalusGame opened the pull request:
:cl:
tweak: made fire colored according to blackbody radiation and rule of cool.
imageadd: some kind of lightning texture
/:cl:

![](https://image.prntscr.com/image/FRAWDEoBQYuv2IwsLmbNjQ.png)

https://i.imgur.com/ETmXnWk.gifv

Looks something like this.
- Makes fire additive blended, which means you can sometimes see through to the lower turf on colder fires
- Makes fire greyscaled using a color matrix and colors it the same way
- Colored fire lets you roughly gauge the temperature in mixing chambers.
- When fire gets "absurdly" hot it has special effect overlays added. (This may not be done very well, please review this)
- Renders fire on top of gas and pretty much everything else, alpha is adjusted for lower temperature fires so you can still see items through the fire.

Things that may still need doing:
- Ensure cutting and adding overlays isn't terrible
- ~~Modify light color without putting a strain on the server (Help appreciated)~~
- Gauss Lerp could be a define
- Code could be cleaned up probably
- ~~Updating the color and effects could be moved to a proc~~

Feedback welcome
Requires #34952 to be merged.

DaedalusGame committed:
- Makes fire colorful and anime

ayyourboibilly commented:
anime

DaedalusGame commented:
should i have put sonic flash instead

Cyberboss commented:
Can you potentially update this less frequently? I'd hate to add this much code to every hotspot process

Cyberboss commented:
Looks good though

Tlaltecuhtli commented:
that picture isnt animated tho

DaedalusGame committed:
- Updates light color on fires
- Makes fire update color less often (1/7th as often)

MoreRobustThanYou commented:
recompile this, lerp should be fixed

WJohn commented:
This is really cool. Only thing to note is that hotter fires don't seem to obscure the floor any more than colder fires.

I'd like to see this "absurdly hot fire effect", too.

MoreRobustThanYou commented:
@DaedalusGame 
```git remote add tg https://github.com/tgstation/tgstation.git
git pull tg master```

Iamgoofball commented:
this does support reaching blue/white heat levels right

DaedalusGame commented:
@WJohn 
![](https://image.prntscr.com/image/v_Vgn8utQTav3x_u1QEkZw.png)
400k Kelvin

![](https://image.prntscr.com/image/rbsDdxLLT76sfdiH78VYjQ.png)
5mil Kelvin (This one feels like it needs to be more violent.

@Iamgoofball I think white is somewhere pretty early like 6000 Kelvin and blue is past that.

banthebantz commented:
You may be interested in the "heat2color" proc found here: https://github.com/ParadiseSS13/Paradise/blob/3eaff16f45583bd60366e96fdbc6bd63f29dc9c4/code/__HELPERS/type2type.dm

Seems a bit better than a bunch of ifs to me.

RemieRichards commented:
@banthebantz he's using them already.

DaedalusGame commented:
The ifs are just for a) putting the old fire texture to use at low temperature and b) making sure that fire above unreasonable temperatures isn't just flat blue (the algorithm is only defined up to 40000 Kelvin)

WJohn commented:
swap out the 5 mil and 400 k and i think we're good.

DaedalusGame commented:
Better idea, I'll change up the math a bit on going purple and make 5mil into rainbow

DaedalusGame committed:
- Changes the transition from blue to purple to be much more gradual, it starts at 40000 and peaks at 120000 Kelvin now
- Changes fusion fire (>4 million Kelvin) to have a slight colorshift to it (It's just the druggy overlay).

DaedalusGame committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into hellburn

DaedalusGame commented:
It's hard man, fire doesn't actually get that hot ingame without fusion and fusion is disabled

DaedalusGame commented:
https://i.imgur.com/FE8NtDC.gifv

DaedalusGame committed:
- Moves fire above gas (and everything else)
- Changes fire alpha for low temperature fires to compensate for items no longer rendering on top of fire

DaedalusGame commented:
As stated in the commit, at request of Joan in coderbus, I've moved fire to a higher layer, to render above gases. Feedback on this change please.

https://i.imgur.com/ETmXnWk.gifv

ShizCalev commented on code/modules/atmospherics/environmental/LINDA_fire.dm:
```diff
@@ -50,7 +50,7 @@
 	icon = 'icons/effects/fire.dmi'
 	icon_state = "1"
-	layer = ABOVE_OPEN_TURF_LAYER
+	layer = FLY_LAYER + 0.1//ABOVE_OPEN_TURF_LAYER
```
RIPPLE_LAYER

DaedalusGame replied:
what the fuck is a hyperspace ripple
nevermind give me a second

DaedalusGame committed:
- Makes the new fire layer a define and makes sure it isn't above "Hyperspace Ripples" whatever that is

ShizCalev commented on code/modules/atmospherics/environmental/LINDA_fire.dm:
```diff
@@ -50,7 +50,7 @@
 	icon = 'icons/effects/fire.dmi'
 	icon_state = "1"
-	layer = FLY_LAYER + 0.1//ABOVE_OPEN_TURF_LAYER
+	layer = FIRE_LAYER //ABOVE_OPEN_TURF_LAYER
```
can remove the comment too, if you wouldn't mind.

DaedalusGame committed:
- Removes comment

DaedalusGame committed:
- FIRE_LAYER is already defined for burning humans.

vuonojenmustaturska commented:
testmerge seems to work, let's do it
